File: BuildUtils/BuildMono.py
import os
import os.path
import subprocess

MONO_COMPILER = "\"C:\\Program Files\\Mono\\bin\\mcs\""
MONO_LIB_PATH = os.path.join("C:\\", "Program Files", "Mono", "lib")

MONO_OPTIONS = ["-target:library"]

MONO_PROJECT_PATH = "MonoScripts"
MONO_AOT_ASSEMBLY_OUTPUT_PATH = os.path.join("MonoScripts", "MonoScripts.dll")

MONO_FILES = [
    "MonoScript.cs",
    "TestScript.cs"
]

MONO_LIBS = []

def build_mono_project(solution_directory):
    args = ["cmd", MONO_COMPILER]
    for option in MONO_OPTIONS:
        args.append(option)
    for mono_script in MONO_FILES:
        mono_script_path = os.path.join(solution_directory, MONO_PROJECT_PATH, mono_script)
        args.append(mono_script_path)
    output_directory = "-out:"+os.path.join(solution_directory, MONO_AOT_ASSEMBLY_OUTPUT_PATH)
    args.append(output_directory)
    print('Running Command: ' + ' '.join(args))
    subprocess.run(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

if __name__ == '__main__':
    cwd = os.getcwd()
    mono_project = os.path.join(cwd, MONO_PROJECT_PATH)
    print('Building Mono Project: ' + mono_project)
    build_mono_project(cwd)
    input("Press Enter to terminate...")

File: BuildUtils/src/build-mono.ts
import * as path from 'path'
import * as fs from 'fs-extra'
import execa from 'execa'
import { MonoBuildConfig } from './data/config-files';

const SDK_ROOT: string = <string>process.env.MONO_SDK_ROOT;
const MONO_COMPILER = path.join(SDK_ROOT, 'bin/mcs')
const MONO_LIB_PATH = path.join(SDK_ROOT, 'lib');

const EXCLUDED_DIRECTORIES = [
    "obj",
    "bin"
];

function getFiles(directory: string, results: string[]) : void {
    fs.readdirSync(directory).forEach((f, index, array) => {
        const fullPath = path.join(directory, f);
        if (fs.statSync(fullPath).isDirectory()
            && EXCLUDED_DIRECTORIES.findIndex((v, i ,o) => { return  v === f; }) === -1) {
            getFiles(fullPath, results);
        } else {
            const extension = fullPath.substring(fullPath.lastIndexOf('.'));
            if (extension === '.cs') {
                results.push(fullPath);
            }
        }
    });
}

export async function BuildMonoProject(monoProject: MonoBuildConfig, rootDirectory: string, buildDirectory: string) {
    console.log(`Building Mono Project: ${monoProject.relativeProjectDirectory}`);
    
    const args: string[] = [];
    args.push(`-target:${monoProject.type}`);
    args.push(`-out:${path.join(buildDirectory, monoProject.relativeTargetDirectory)}`);

    const fullPath = path.join(rootDirectory, monoProject.relativeProjectDirectory);
    const allFiles: string[] = [];
    getFiles(fullPath, allFiles);
    for (const monoScriptPath of allFiles) {
        console.log(monoScriptPath);
        args.push(monoScriptPath);
    }
    
    const exec = execa(MONO_COMPILER, args);
    exec.stdout.on('data', data => console.log(data.toString()));
    exec.stderr.on('data', data => console.log(data.toString()));
    await exec;
}
